### ISSUE TITLE
Error during the dump

### ISSUE THREAD

**Issue by DarkerUltor** [DarkMatterCore/gcdumptool#11]
i was dumping super mario odissey, when the dump is at 53% i get that error: Failed to write chunk to offset 0x00000000FFF18000. i tried with another dumpers, but i get the same error, is it the cart? im on 6.2.0

**Comment by LilPuff420** [comment]
Yo, us anyone or the owner of this repo gonna help? I have the same error stuck at the same percentage which is 53 and I tried it with let's go Pikachu.

**Comment by DarkMatterCore** [comment]
@DarkerUltor

Hi there. Sorry for taking so long to answer, a lot of things have happened in my personal life in the past few months, and they have kept me busy enough to stop browsing the Internet as much as I'd like.

Given that other dumpers also fail to dump the same chunk, it may very well be some kind of gamecard malfunction. However, unless I see either an screenshot with the error code provided (or the error code itself), it'll be hard for me to debug this particular problem.

Nonetheless, feel free to test the attached test build.

[gcdumptool_test_20190422.zip](https://github.com/DarkMatterCore/gcdumptool/files/3104155/gcdumptool_test_20190422.zip)

**Comment by Hking0036** [comment]
I am experiencing the same problem, but with Pokemon: Let's Go, Eevee! and not Mario. I have dumped by SMO Rev 0 Cart before with an older version (1.0.3) of GCDumpTool, and didn't have any problems. This might be a problem with newer carts.

I was attempting to dump Pokemon on GCDumpTool version 1.0.7 and this is the error code that GCDumpTool produced: 
![photo_2019-04-27_12-07-23](https://user-images.githubusercontent.com/1592179/56852131-0b6cf400-68e5-11e9-85d6-b1fcab2ce019.jpg)

**Comment by DarkMatterCore** [comment]
@Hking0036 Thanks for your input on this matter. It is greatly appreciated.

Weird that the bug is reproduced when the split option is disabled, since file I/O is more complicated when it's actually enabled. A single fwrite() call is performed per each loop iteration in that case, so it must be happening here for some kind of reason:

https://github.com/DarkMatterCore/gcdumptool/blob/fd4b7e933b57d76c53d8d931cb6a7b95ba4e11b7/source/dumper.c#L583-L589

I'll look into it ASAP. Would you be able to try out any test builds?

**Comment by Hking0036** [comment]
I can try out test builds, but unfortunately I'm away from home until tomorrow, so short-term I won't really be able to. In regards to the setup:

* SD Card is a Samsung brand EVO 128 GB card (the yellow one), it has one big old FAT32 partition.
* FW is 8.0.1, I'm running Atmosphere 0.8.9, using KosmosNX to load it.
* For this instance, I ran it as an applet with the album.

In regards to the SD checking, it looks like you might be able to use `GetSdCardUserAreaSize()` to check for < 34359738368 bytes of usable space and auto-toggle it on but obviously that won't necessarily catch all use cases, and is less than optimal.

**Comment by DarkMatterCore** [comment]
@Hking0036 Don't worry about it, just getting any help on this is good enough. Doesn't matter if its slow or not. :)

If you're using FAT32, that explains why the error is happening: the application is exceeding the `4 GiB - 1` file size limit imposed by FAT32 exactly at offset `0xFFF18000`. A buffer size of 1 MiB is used, so `0x0FFF18000 + 0x000100000 = 0x100018000` (the limit itself being `0x0FFFFFFFF`).

So, this would just be related to the fact the application actually needs the user to choose if they're using FAT32 or exFAT. We \*really\* need to add a way to check this automatically.

I have also considered using `GetSdCardUserAreaSize()` in the past and rapidly scrapped that idea. `IsExFatSupported()` actually checks if the exFAT-compatible PRODINFO update has been installed (but you could very well be using a FAT32 partition anyway), so it's not useful.

I've been looking for a way to gain raw access to the SD card storage device in order to manually check the partition signature, but I've had no luck. This is actually how gamecard dumping works, by gaining raw access to it. Without an IPC call to achieve this, I suppose it's gonna be difficult.

I already have an internal v1.0.8 build with many code style changes, and better error reporting for `fwrite()` is one of them. I'll upload the file shortly so you can test it tomorrow (I'm on my phone right now).

**Comment by Hking0036** [comment]
Here's what I got out of your test build:
![2019042723070000-DB1426D1DFD034027CECDE9C2DD914B8](https://user-images.githubusercontent.com/1592179/56858039-3fc2cd80-6942-11e9-8072-24f4a66e57f4.jpg)
![2019042723131700-DB1426D1DFD034027CECDE9C2DD914B8](https://user-images.githubusercontent.com/1592179/56858040-40f3fa80-6942-11e9-8093-fdf5cad2f825.jpg)

**Comment by DarkMatterCore** [comment]
@Hking0036 Yup, just as I thought: the problem is definitely related to the FAT32 file size limit. That's why it works when you enable the split option. The displayed information matches up with the sum I wrote in my previous comment.

Unless a solution for detecting exFAT partitions is found, the application will still rely on the user to determine this, sadly.

**Comment by sttng** [comment]
Wouldn't it help to look into Hekate and check how it does Fat32 / ExFat handling (https://github.com/CTCaer/hekate/tree/master/bootloader/libs/fatfs)

**Comment by DarkMatterCore** [comment]
@sttng Nope, it wouldn't. Hekate has full hardware access because it runs on a pretty early boot stage, so it offers features that simply can't be done while running under a homebrew context.

CFWs need to be modified to provide some sort of custom IPC call that simply reports the SD card partition type that was detected while booting.

**Comment by sttng** [comment]
Why not keep track of how many bytes are written, and in case there is an error at 4 GiB, catch that error and start split - in case of no error at 4 GiB just continue writing ? There may be a rare case of split because of an actual sector problem at exactly 4 GiB while being on exfat I guess (also triggering a 'false' write error). But that should be rather low.

**Comment by DarkMatterCore** [comment]
@sttng Technically, the error is already being catched, but nothing more than displaying a message that suggests the user to enable the FAT32 option is being done at the moment.

If I have enough free time, I'll try to look into that.

**Comment by DarkMatterCore** [comment]
The ongoing rewrite will generate output dumps using ConcatenationFiles if they exceed the 4 GiB mark, even under exFAT. This will remove the need for a FAT32 option.
